setup-env

```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
-fi
-if [ -e "$BDIR" ]; then
-    BDIR=`readlink -f "$BDIR"`
-    if [ ! -d "$BDIR" ]; then
-	echo "Usage: setup-env [dirname]"
-	return 1
+
+usage()
+{
+    echo -e "
+Usage: MACHINE=<machine> DISTRO=<distro> source $PROGNAME <build-dir>
+Usage:                                   source $PROGNAME <build-dir>
+    <machine>    machine name
+    <distro>     distro name
+    <build-dir>  build directory
+The first usage is for creating a new build directory. In this case, the
+script creates the build directory <build-dir>, configures it for the
+specified <machine> and <distro>, and prepares the calling shell for running
+bitbake on the build directory.
+The second usage is for using an existing build directory. In this case,
+the script prepares the calling shell for running bitbake on the build
+directory <build-dir>. The build directory configuration is unchanged.
+"
+
+    ls layers/*/conf/machine/*.conf > /dev/null 2>&1
+    ls layers/meta-testdistro/conf/distro/*.conf > /dev/null 2>&1
+    if [ $? -eq 0 ]; then
+        echo -e "
+Supported machines: `echo; ls layers/*/conf/machine/*.conf \
+| sed s/\.conf//g | sed -r 's/^.+\///' | xargs -I% echo -e "\t%"`
+Supported Tegra test distros: `echo; ls layers/meta-testdistro/conf/distro/*.conf \
+| sed s/\.conf//g | sed -r 's/^.+\///' | xargs -I% echo -e "\t%"`
+Examples:
+- To create a new Yocto build directory:
+  $ MACHINE=jetson-tx2 DISTRO=testdistro source $PROGNAME build
+- To use an existing Yocto build directory:
+  $ source $PROGNAME build
+"
     fi
+}
+
+clean_up()
+{
+   unset LIST_MACHINES VALID_MACHINE
+   unset CWD TEMPLATES SHORTOPTS LONGOPTS ARGS PROGNAME
+   unset generated_config updated
+   unset MACHINE SDKMACHINE DISTRO OEROOT TEMPLATECONF BBPATH
+}
+
+# get command line options
+SHORTOPTS="h"
+LONGOPTS="help"
+
+ARGS=$(getopt --options $SHORTOPTS  \
+  --longoptions $LONGOPTS --name $PROGNAME -- "$@" )
+# Print the usage menu if invalid options are specified
+if [ $? != 0 -o $# -lt 1 ]; then
+   usage && clean_up
+   return 1
+fi
+
+eval set -- "$ARGS"
+while true;
+do
+    case $1 in
+        -h|--help)
+           usage
+           clean_up
+           return 0
+           ;;
+        --)
+           shift
+           break
+           ;;
+    esac
+done
+
+if [ "$(whoami)" = "root" ]; then
+    echo "ERROR: do not use the BSP as root. Exiting..."
+fi
+
+if [ ! -e $1/conf/local.conf ]; then
+    build_dir_setup_enabled="true"
 else
-    mkdir -p "$BDIR"
-    BDIR=`readlink -f "$BDIR"`
+    build_dir_setup_enabled="false"
+fi
+
+if [ "$build_dir_setup_enabled" = "true" ] && [ -z "$MACHINE" ]; then
+    usage
+    echo -e "ERROR: You must set MACHINE when creating a new build directory."
+    clean_up
+    return 1
 fi
+
+if [ -z "$SDKMACHINE" ]; then
+    SDKMACHINE='x86_64'
+fi
+
+if [ "$build_dir_setup_enabled" = "true" ] && [ -z "$DISTRO" ]; then
+    usage
+    echo -e "ERROR: You must set DISTRO when creating a new build directory."
+    clean_up
+    return 1
+fi
+
```

Comment:
Suggested addition here:
```
(cd $CWD; git submodule sync > /dev/null 2>&1)
(cd $CWD; git submodule status | grep '^-' | awk '{print $2}' | xargs git submodule update --init --)
```
One common problem I've run into with my submodule-based distros is people forgetting to initialize the submodules on first clone, or forgetting to update the submodules after a `git pull`.   I never got round to adding that here, though, because I never forget. :)